libraries/cmd/dock_utils.cpp

```diff
@@ -70,103 +74,126 @@ int CanDock(Unit *dock, Unit *ship, bool ignore_occupancy) {
         return -1;
     }
 
+    const bool is_player_starship = _Universe->isPlayerStarship(ship) || _Universe->isPlayerStarship(dock);
 
     double range = DistanceTwoTargets(dock, ship);
 
     // Planet Code
     if (dock->isUnit() == Vega_UnitType::planet) {
         range -= dock->rSize() * (configuration()->dock.dock_planet_radius_percent - 1);
         if (range < 0) {
+            if (is_player_starship) {
+                VS_LOG(trace, "Planet. Within range. CanDock = 0");
+            }
             return 0;
         } else {
+            if (is_player_starship) {
+                VS_LOG(trace, "Planet. Out of range. CanDock = -1");
+            }
             return -1;
         }
     }
 
 
     if(configuration()->dock.simple_dock) {
         range = DistanceTwoTargets(dock, ship);
+        if (is_player_starship) {
+            VS_LOG(trace, (boost::format("CanDock: simple_dock is true. range = %1%") % range));
+        }
 
         if (range < configuration()->dock.simple_dock_range) {
+            if (is_player_starship) {
+                VS_LOG(trace, "simple_dock is true. Within range. CanDock = 0");
+            }
             return 0;
         } else {
-            return -1;
-        }
-    }
-
-    if (ignore_occupancy) {
-        for (unsigned int i = 0; i < dock->pImage->dockingports.size(); ++i) {
-            if (!dock->pImage->dockingports[i].IsOccupied()) {
-                return i;
+            if (is_player_starship) {
+                VS_LOG(trace, "simple_dock is true. Out of range. CanDock = -1");
             }
+            return -1;
         }
     }
 
     //don't need to check relation: already cleared.
 
     // If your unit has docking ports then we check if any of our docking
     // ports overlap with any of the station's docking ports.
-    // Otherwise we simply check if our unit overlaps with any of the
+    // Otherwise, we simply check if our unit overlaps with any of the
     // station's docking ports.
     for (unsigned int i = 0; i < dock->pImage->dockingports.size(); ++i) {
-        if (!dock->pImage->dockingports.empty()) {
-            for (unsigned int j = 0; j < dock->pImage->dockingports.size(); ++j) {
+        if (!ship->pImage->dockingports.empty()) {
```

Comment:
Without understanding anything about the code, the `ship` doesn't have `dockingports`. The `dock` does.

Reply:
This block of code is actually supposed to handle the case where both the ship AND the dock have docking ports. Best I can tell.

Reply:
We can handle that as a follow-up issue and PR.

Reply:
I think what the devs were aiming for is to only dock if the dock of the ship collides with the dock of the station.
Given that we're comparing ship dock to ship dock, this will **always** work.